asapdiscovery/simulation/scripts/MProFunction.py

```diff
@@ -0,0 +1,239 @@
+
+
+def simulate_mpro(input_dir,protein,ligand,nsteps,output_dir):
+    ## Necessary for Reporters
+    save = re.search(r"Mpro-P+\d{4}", protein).group(0)
+    logfile = output_dir / f"{save}-log.txt"
+    logger = logging.getLogger(save)
+    logger.setLevel(logging.DEBUG)
+    handler = logging.FileHandler(str(logfile), mode="w")
+    handler.setLevel(logging.DEBUG)
+    formatter = logging.Formatter("%(asctime)s | %(levelname)s | %(message)s")
+    handler.setFormatter(formatter)
+    logger.addHandler(handler)
+
+    num_steps = int(nsteps) # number of integrator steps
+    n_snapshots = int(num_steps / reportingInterval) # calculate number of snapshots that will be generated
+    num_steps = n_snapshots * reportingInterval # recalculate number of steps to run
+
+    # Read the molfile into RDKit, add Hs and create an openforcefield Molecule object
+    logger.info(':pill:  Reading ligand')
+    rdkitmol = Chem.SDMolSupplier(ligand)[0]
+    logger.info(f':mage:  Adding hydrogens')
+    rdkitmolh = Chem.AddHs(rdkitmol, addCoords=True)
+    # ensure the chiral centers are all defined
+    Chem.AssignAtomChiralTagsFromStructure(rdkitmolh)
+    ligand_mol = Molecule(rdkitmolh)
+
+    # Initialize a SystemGenerator
+    logger.info(':factory: System Combination')
+    forcefield_kwargs = {'constraints': app.HBonds, 'rigidWater': True, 'removeCMMotion': False, 'hydrogenMass': 4*unit.amu }
+    periodic_forcefield_kwargs = {'nonbondedMethod': app.PME}
+    system_generator = SystemGenerator(
+        forcefields=['amber/ff14SB.xml', 'amber/tip3p_standard.xml'],
+        small_molecule_forcefield='openff-1.3.1',
+        molecules=[ligand_mol], cache='cache.json',
+        forcefield_kwargs=forcefield_kwargs, periodic_forcefield_kwargs=periodic_forcefield_kwargs)
+
+    # Use Modeller to combine the protein and ligand into a complex
+    logger.info(':cut_of_meat:  Reading protein')
+    protein_pdb = PDBFile(protein)
+    logger.info(':sandwich:  Preparing complex')
+    modeller = Modeller(protein_pdb.topology, protein_pdb.positions)
+    # This next bit is black magic.
+    # Modeller needs topology and positions. Lots of trial and error found that this is what works to get these from
+    # an openforcefield Molecule object that was created from a RDKit molecule.
+    # The topology part is described in the openforcefield API but the positions part grabs the first (and only)
+    # conformer and passes it to Modeller. It works. Don't ask why!
+    modeller.add(ligand_mol.to_topology().to_openmm(), ligand_mol.conformers[0].to_openmm())
+
+    # We need to temporarily create a Context in order to identify molecules for adding virtual bonds
+    logger.info(f':microscope:  Identifying molecules')
+    integrator = openmm.VerletIntegrator(1*unit.femtoseconds)
+    system = system_generator.create_system(modeller.topology, molecules=ligand_mol)
+    context = openmm.Context(system, integrator, openmm.Platform.getPlatformByName('CUDA'))
```

Comment:
you should use your platform variable here

Reply:
Got it!